### PR TITLE
Instant::checked_duration_since

### PR DESCRIPTION
**Pull request by vi**

### CODE REVIEW

**Comment by Centril** [comment]
cc https://github.com/rust-lang/rust/issues/58402

**Comment by vi** [comment]
How tests are supposed to be added for fresh `#[unstable]` items?

**Comment by Centril** [comment]
@vi You add `#![feature(checked_duration_since)]` to the crate root of libstd.

File: src/libstd/time.rs [hunk 1]
```diff
@@ -218,6 +218,30 @@ impl Instant {
         self.0.sub_instant(&earlier.0)
     }
 
+    /// Returns the amount of time elapsed from another instant to this one,
+    /// or None if that instant is earlier than this one.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use std::time::{Duration, Instant};
```
Comment (Centril) [inline src/libstd/time.rs]:
```suggestion
    /// #![feature(checked_duration_since)]
    /// use std::time::{Duration, Instant};
```

File: src/libstd/time.rs [hunk 2]
```diff
@@ -626,6 +650,12 @@ mod tests {
         (a - Duration::new(1, 0)).duration_since(a);
     }
 
+    #[test]
+    fn checked_instant_duration_nopanic() {
+        let a = Instant::now();
+        (a - Duration::new(1, 0)).checked_duration_since(a);
```
Comment (scottmcm) [inline src/libstd/time.rs]: nit: probably worth checking that it's None, too.

**Comment by scottmcm** [comment]
Random thought: would `saturating_duration_since` make sense too?

**Comment by vi** [comment]
@scottmcm I was also thinking about returning `Result<Duration,Duration>`, but decided to keep it the same as with other `checked_*`.

I'll probably add `saturating_duration_since`, returning zero duration instead of None.

Is wrong order of `Instant`s the only overflow situation possible in `duration_since` or it can also happen if two `Instant`s are too far apart?

**Review by dtolnay — COMMENTED** [review]
Could you explain more about the concrete use case for this? The usual use of Instant is for measuring duration between a "start" and "end" instant, in which you always know which one is first.

I think #55527 was well justified:

> Since `SystemTime` is opaque there is no way to check if the result of an addition will be in bounds. That makes the `Add<Duration>` trait completely unusable with untrusted data. This is a big problem because adding a `Duration` to `UNIX_EPOCH` is the standard way of constructing a `SystemTime` from a unix timestamp.

but there doesn't seem to be a similar compelling justification in #58402.

**Comment by dtolnay** [comment]
> Another idea, for laziers who don't want to remember which one is earlier and which one is later

I agree that these are things we could add, but I would prefer not to pack the standard library with public signatures that will only ever get called by the test suite of the standard library. I would like for anything we add to be justified more strongly than this.

**Comment by vi** [comment]
The reason is similar to `checked_add` case: untrusted input.

Plain `Instant::duration_since` is OK when both Instants come from `Instant::now`, but when one or both instants are actually some baseline `Instant`, offseted by a constructed `Duration` based on untrusted input, order becomes less reliable, hence the need for checked and saturating version.

**Comment by dtolnay** [comment]
Hmm, I am still not seeing it. What I question is not whether it's possible to provoke a panic from duration_since, but whether there would ever be code that is expressed most clearly using checked_duration_since / saturating_duration_since rather than some other way. Would you personally want to use either of these methods today in existing code if they were available and stable? Separately, would you be able to point out some existing real-world code where these methods are the clearest solution?

**Comment by vi** [comment]
Most recent [example](duration_since) from my code:

> `let rtt4 = if now >= send_time { now - send_time } else { Duration::from_secs(999)};`

`now.saturating_duration_since(send_time)` would probably be OK for this, or a `checked_duration_since` with more explicit fallback.

---

Searching Github for `duration_since` in Rust code leads mostly to [`SystemTime::duration_since`](https://doc.rust-lang.org/std/time/struct.SystemTime.html#method.duration_since) which returns `Result`, so is more like a `checked_duration_since`.

**Comment by dtolnay** [comment]
From your code:

```rust
let now = Instant::now();
if now > deadline {
    break;
}

let s: u64 = /* untrusted */;
let ns: u32 = /* untrusted */;

let since_base = Duration::new(s, ns);
let send_time = time_base + since_base;
let rtt4 = if now >= send_time { now - send_time } else { Duration::from_secs(999) };
```

This wouldn't be entirely fixed by checked_duration_since because the `Duration::new` and the `time_base + since_base` can both overflow and panic. I might write this securely as:

```rust
let s = Duration::from_secs(/* untrusted */);
let ns = Duration::from_nanos(/* untrusted */);

let to_send = s.checked_add(ns).unwrap_or(...);
let to_now = now - time_base;
let rtt4 = to_now.checked_sub(to_send).unwrap_or(...);
```

Maybe this suggests a best practice of sticking as much as possible to `Instant::elapsed` + subtraction of a trusted instant, and doing untrusted work only in terms of Duration which has a less dangerous API. What do you think?

Thanks for the example use case -- I can see from this code that Instant could benefit from a carefully designed checked API.

I think there is also room for a crate called `easytime` or something like that which provides Instant and Duration wrapper types with all the Add/Sub impls but that never panic, instead keeping track of whether overflow occurred. This would make the obvious code also correct without having to be careful about where all the checked_subs go:

```rust
use easytime::{Duration, Instant};


let s: u64 = /* untrusted */;
let ns: u32 = /* untrusted */;

let since_base = Duration::new(s, ns);
let rtt4 = now - (time_base + since_base);

// some method to go from easytime::Duration to std::time::Duration:
let rtt4 = rtt4.unwrap_or(...);
```

**Review by dtolnay — APPROVED** [review]
I think these methods are good. Sorry for giving you grief! I copied your rationale over to the tracking issue.

I would still be interested in seeing an easytime crate to make this stuff easier to write correctly.

**Comment by vi** [comment]
> Maybe this suggests a best practice of sticking as much as possible to Instant::elapsed + subtraction of a trusted instant, and doing untrusted work only in terms of Duration which has a less dangerous API. What do you think?

I want to reason in terms of moments of time and timespans. Although it is obviously possible to use `Durations` as moments of time (by always keeping in mind the base), it loses typestate protection against adding two `Instant`s together and also keeps code less self-documenting (i.e. `/// This Duration is actually a moment of time, not a duration of something`).

**Comment by scottmcm** [comment]
> I would still be interested in seeing an easytime crate to make this stuff easier to write correctly.

Hmm, we have `num::Wrapping` to call `wrapping_*` methods.  What's your first instinct about a `num::Checked` to call all the `checked_*` methods?  (I was thinking it'd be a newtype over an option.)

**Comment by dtolnay** [comment]
@scottmcm:

> What's your first instinct about a `num::Checked` to call all the `checked_*` methods?

I think it would be less good than https://github.com/taiki-e/easytime.

I think construction of `Checked` would be confusing or complicated:

```rust
// not panic-free
let _ = Checked::new(Duration::new(s, ns));

// weird and not panic-free
let _ = Checked(Some(Duration::new(s, ns)));

// :( less important words up front
let _ = Checked::new_duration(s, ns);

// ???
let _ = Checked::now();
let _ = Checked::<Instant>::now();
let _ = Checked::now_instant();
let _ = Checked::instant_now();

// checked_ prefix implies it returns Option, not Checked<Duration>
let _ = Duration::checked_new(s, ns);

// not bad...
let _ = Duration::new_checked(s, ns);

// ... but for Instant?
let _ = Instant::now_checked(); // ???

// easytime
let _ = Duration::new(s, ns);
let _ = Instant::now();
```

**Pull request merged**